File: OrdoRubim.rb
# ════════════════════════════════════════════════════════════════════════════
#  ORDO RUBIM — SYMPHONIA CODICIS (RUBY VERSION)
#  LIBRETTO: MATERIA OBSCURA
#  INTERPRETATIO: MRI (Matz's Ruby Interpreter)
# ════════════════════════════════════════════════════════════════════════════

require 'singleton' # UNUS (Один)
require 'thread'    # FILUM (Нить)

# [ACT I: VOCATIO ELEMENTORUM]

# ─── CANTUS PRIMUS: CHAOS ET ORDO ────────────

module Chaos
  # // Hic sunt dracones.
  class RawInput
    attr_reader :data

    def initialize(data)
      # // Verbum congelatum est.
      @data = data.to_s.freeze 
    end

    # // Nihil ex nihilo fit.
    def void?
      @data.nil? || @data.strip.empty?
    end
  end

  module EntropyFilter
    module_function

    # // Purgamentum init, purgamentum exit.
    def purge(input)
      # // Si silentium est...
      if input.void?
        # // ...clamor erroris audietur!
        raise StandardError, "ABYSSUS: Vacuitas non accipitur."
      end
    end
  end
end

# ─── CANTUS SECUNDUS: LEX AETERNA ──────────────

module Dictatura
  module Accuratio
    module_function

    # // Mensura sine errore.
    def verify_purity(value)
      # // Numerus insanus vetatur.
      if value.nan? || value.infinite?
        raise RangeError, "HERESIS: Numerus non existit."
      end
    end

    # // Quod imperfectum est, corrigetur.
    def snap_to_grid(value)
      limen = 1e-6 
      return 0.0 if value.abs < limen
      value
    end
  end

  class ImmutableLaw
    # // Unus rex, una lex.
    @gate_keeper = Mutex.new

    def self.acquire_control
      # // Ianua clauditur.
      @gate_keeper.synchronize do
        # // Intra sanctum.
        yield
      end
      # // Ianua aperitur.
    end
  end
end

# ─── CANTUS TERTIUS: MACHINATIO ────────────────────

module MateriaObscura
  module VoidEngine
    module_function

    # // Et verbum caro factum est.
    def invoke(will)
      # // Machina obsequitur.
      puts "[RUBIM] >>> Fiat: #{will}"
    end
  end
end

# ─── CANTUS QUARTUS: MAGISTER ────────────────────

class Magister
  include Singleton

  # // Ego sum Alpha et Omega.

  # // In manu mea potestas est. 
  def dominari_materia(intentio)
    
    # // 1. INTROITUS (Вход)
    input = Chaos::RawInput.new(intentio)

    # // Si indignus es, discede. 
    Chaos::EntropyFilter.purge(input)

    # // 2. SILENTIUM 
    # // Tempus sistit. 
    Dictatura::ImmutableLaw.acquire_control do
      
      # // 3. ACTIO 
      # // Mundus mutatur. 
      MateriaObscura::VoidEngine.invoke(intentio)

    end # // 4. EXPLICIT (Mutex releases here)
  end
end

# ─── CANTUS QUINTUS: ADYTUM ────────────────────────

module Initiatus
  def self.included(base)
    # // Nomen est omen. (Имя есть знак.)
    unless base.const_defined?(:NOMEN_UMBRAE)
      raise NameError, "DAMNATIO: Classis #{base} sine nomine est."
    end
    
    nomen = base.const_get(:NOMEN_UMBRAE)
    if nomen.nil? || nomen.strip.empty?
       raise StandardError, "DAMNATIO: Nomen vacuum."
    end
  end

  # // Opus perfectum requiritur. 
  def opus_magnum(guru)
    raise NotImplementedError, "HERESIS: Opus Magnum non completum est."
  end
end

class Adytum
  def self.open_gates(adept_class)
    # // Nomina sunt odiosa? Non hic. 
    unless adept_class.include?(Initiatus)
      raise TypeError, "PROFANUS: Classis non initiata est."
    end

    nomen = adept_class.const_get(:NOMEN_UMBRAE)

    puts "\n>>> GLORIA IN EXCELSIS CODEX."
    puts ">>> Adventus: #{nomen}.\n\n"

    begin
      guru = Magister.instance
      neophyte = adept_class.new
      
      # // Incipit Tragoedia. 
      neophyte.opus_magnum(guru)
      
      puts "\n>>> CONSUMMATUM EST. (Свершилось.)"

    rescue => e
      # // Dies irae, dies illa. 
      STDERR.puts "\n!!! LACRIMOSA: #{e.message}"
      # Ruby way to exit with error code
      exit(0xDEAD) 
    end
  end
end

# ─── MACRO ACTIVATIONIS ────────────────────────────

def in_umbra_igitur_pugnabimus(adept_class)
  Adytum.open_gates(adept_class)
end

# ════════════════════════════════════════════════════════════════════════════
# EXEMPLUM
# ════════════════════════════════════════════════════════════════════════════

=begin

class Architectus
  include Initiatus
  
  NOMEN_UMBRAE = "Architect_Ruby"

  def opus_magnum(guru)
    guru.dominari_materia("Fiat Lux")
  end
end

in_umbra_igitur_pugnabimus(Architectus)

=end

File: OrdoRubim.ts
/**
 * ════════════════════════════════════════════════════════════════════════════
 * ORDO RUBIM — SYMPHONIA CODICIS (TYPESCRIPT VERSION)
 * LIBRETTO: MATERIA OBSCURA
 * TARGET: DENO / NODE / BUN (SERVER-SIDE OR STRICT CLIENT)
 * ════════════════════════════════════════════════════════════════════════════
 */

// [ACT I: VOCATIO ELEMENTORUM]

// ─── CANTUS PRIMUS: CHAOS ET ORDO ───────────────────────────────────────────

namespace Chaos {
    // // Hic sunt dracones.
    export class RawInput {
        // // Verbum congelatum est.
        constructor(public readonly data: string) {}

        // // Nihil ex nihilo fit.
        public isVoid(): boolean {
            return !this.data || this.data.trim().length === 0;
        }
    }

    export class EntropyFilter {
        // // Purgamentum init, purgamentum exit.
        public static purge(input: RawInput): void {
            // // Si silentium est...
            if (input.isVoid()) {
                // // ...clamor erroris audietur!
                throw new Error("ABYSSUS: Vacuitas non accipitur.");
            }
        }
    }
}

// ─── CANTUS SECUNDUS: LEX AETERNA ───────────────────────────────────────────

namespace Dictatura {
    export class Accuratio {
        // // Mensura sine errore.
        public static verifyPurity(value: number): void {
            // // Numerus insanus vetatur.
            if (Number.isNaN(value) || !Number.isFinite(value)) {
                throw new Error("HERESIS: Numerus non existit.");
            }
        }

        // // Quod imperfectum est, corrigetur.
        public static snapToGrid(value: number): number {
            const LIMEN = 1e-6; // Порог тишины
            if (Math.abs(value) < LIMEN) return 0.0;
            return value;
        }
    }

    // Имитация Мьютекса для однопоточного Event Loop 
    export class ImmutableLaw {
        private static _isLocked = false;

        // // Unus rex, una lex.
        public static acquireControl(): { release: () => void } {
            if (this._isLocked) {
                // В реальном async коде здесь был бы await queue
                throw new Error("VIOLATIO: Realitas jam occupata est (Re-entry detected).");
            }
            
            // // Ianua clauditur.
            this._isLocked = true;

            // Возвращаем "Ключ" (Disposable pattern)
            return {
                release: () => {
                    // // Ianua aperitur.
                    this._isLocked = false;
                }
            };
        }
    }
}

// ─── CANTUS TERTIUS: MACHINATIO ─────────────────────────────────────────────

namespace MateriaObscura {
    export class VoidEngine {
        // // Et verbum caro factum est.
        public static invoke(will: string): void {
            // // Machina obsequitur.
            console.log(`[RUBIM] >>> Fiat: ${will}`);
        }
    }
}

// ─── CANTUS QUARTUS: MAGISTER ───────────────────────────────────────────────

export class Magister {
    // Singleton instance holder
    private static _instance: Magister;

    // Private constructor to prevent instantiation outside
    private constructor() {}

    // // Ego sum Alpha et Omega.
    public static get instance(): Magister {
        if (!this._instance) {
            this._instance = new Magister();
        }
        return this._instance;
    }

    // // In manu mea potestas est. 
    public dominariMateria(intentio: string): void {
        
        // // 1. INTROITUS (Вход)
        const input = new Chaos.RawInput(intentio);

        // // Si indignus es, discede. 
        Chaos.EntropyFilter.purge(input);

        // // 2. SILENTIUM 
        // // Tempus sistit. 
        const guard = Dictatura.ImmutableLaw.acquireControl();

        try {
            // // 3. ACTIO 
            // // Mundus mutatur. 
            MateriaObscura.VoidEngine.invoke(intentio);
        } finally {
            // // 4. EXPLICIT 
            // // Pax vobiscum. 
            guard.release();
        }
    }
}

// ─── CANTUS QUINTUS: ADYTUM ─────────────────────────────────────────────────

export interface IInitiatus {
    // // Opus perfectum requiritur. 
    opusMagnum(guru: Magister): void;
}

export interface IInitiatusConstructor {
    // // Nomen est omen. (Имя есть знак.)
    readonly NOMEN_UMBRAE: string;
    new (): IInitiatus;
}

export class Adytum {
    public static openGates(AdeptClass: IInitiatusConstructor): void {
        // // Nomina sunt odiosa? Non hic. 
        const nomen = AdeptClass.NOMEN_UMBRAE;

        // // Sine nomine, sine vita. 
        if (!nomen || nomen.trim().length === 0) {
             throw new Error("DAMNATIO: Nomen vacuum.");
        }

        console.log("\n>>> GLORIA IN EXCELSIS CODEX.");
        console.log(`>>> Adventus: ${nomen}.\n`);

        try {
            const guru = Magister.instance;
            const neophyte = new AdeptClass();
            
            // // Incipit Tragoedia. 
            neophyte.opusMagnum(guru);
            
            console.log("\n>>> CONSUMMATUM EST. (Свершилось.)");

        } catch (e) {
            // // Dies irae, dies illa. 
            const msg = e instanceof Error ? e.message : String(e);
            console.error(`\n!!! LACRIMOSA: ${msg}`);
            process.exit(0xBADC0DE);
        }
    }
}

// ─── MACRO ACTIVATIONIS ─────────────────────────────────────────────────────

// Функция-обертка для запуска
export function InUmbraIgiturPugnabimus(AdeptClass: IInitiatusConstructor) {
    Adytum.openGates(AdeptClass);
}

// ════════════════════════════════════════════════════════════════════════════
// EXEMPLUM 
// ════════════════════════════════════════════════════════════════════════════

/*
class Architectus implements IInitiatus {
    static readonly NOMEN_UMBRAE = "Architect_TS";

    opusMagnum(guru: Magister): void {
        guru.dominariMateria("Fiat Lux");
    }
}

InUmbraIgiturPugnabimus(Architectus);
*/